Add extension impact script

sgraband opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Adds a script that logs a csv table with an overview of the impact an extension has on Theia's startup time. The script compares the startup time of a defined base application `scripts/performance/base-package.json` with the base application + an additional extension.
The performance measurement script from #(link to performance measurement PR) is used to measure the startup time. Therefore this PR depends on that PR. The script can receive multiple extensions, which are then tested one by one. If no extensions are provided all extensions in the `packages` folder will be tested.

Via parameters you can set a 
- baseTime for a previous measurement of the base application
- list of extensions that shall be tested
- number of runs for each measurement
- For more information see the README in `scripts/performance`

Example Result:

|Extension name|MEAN (10 runs) (in s)|STDEV (in ms)|CV (%)|DIFF (in ms)|
|---|---|---|---|---|
|Base theia|2.027|0.084|4.144|-|
|"@theia/git":"1.17.0"|2.103|0.041|1.950|0.076|

Contributed on behalf of STMicroelectronics

Signed-off-by: Simon Graband <sgraband@eclipsesource.com>
#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Run `node ./extension-impact.js` from the `scripts/performance` directory. And have a look at the resulting table. 
Note: This may take a while. To reduce the time either set the runs to a smaller number than 10 or use only a set of extensions.
#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

cdamus reviewed (commented):
# Code Review (Theia, ST)

This works well when I run it on specific extensions. Very nice! I have very minor comments on the code.

If, for whatever reason, the developer cancels execution of the script, it leaves the `examples/browser/package.json` file changed, which the developer then could accidentally stage and include in a commit, breaking the example. Can all of the manipulation be done instead on a clean copy of the example? I can imagine it would be routine to cancel the script after some big anomaly is printed out, or because all of the interesting output is already written, etc.

A question: if I am concerned about the interaction of the Theia workspace with extension start-up performance (for example, performance dependent on number of files in the workspace or the presence/content of specific files), how can I account for that? Would it be useful to include an option to specify a workspace to open?

Finally, when running the test on all of the packages in the monorepo, I see a minor issue with error handling. When the script fails to get a measurement (usually because it failed to launch the browser, it seems), then the report outputs a line of NaNs for the measurements (see snippet, below). Perhaps this case could be better handled by outputting an explicit error row to the table instead of the NaNs that assumed valid inputs to the calculation. Happily, the error traces are written to `stderr` and so they don't interfere with the redirection of the report to a CSV file.

```
 node ./extension-impact.js --runs 7                              
Extension name, MEAN (7 runs) (in s), STDEV (in ms), CV (%), DIFF (in ms)
Base theia, 2.030, 0.009, 0.443, -
"@theia/bulk-edit": "1.17.0", 2.050, 0.010, 0.488, 0.020
"@theia/callhierarchy": "1.17.0", 2.051, 0.016, 0.780, 0.021
"@theia/console": "1.17.0", 2.049, 0.022, 1.074, 0.019
"@theia/debug": "1.17.0", 2.021, 0.017, 0.841, -0.009
"@theia/editor": "1.17.0", 2.043, 0.013, 0.636, 0.013
"@theia/editor-preview": "1.17.0", 2.041, 0.025, 1.225, 0.011
"@theia/external-terminal": "1.17.0", 2.017, 0.014, 0.694, -0.013
"@theia/file-search": "1.17.0", 2.025, 0.016, 0.790, -0.005
"@theia/filesystem": "1.17.0", 2.041, 0.038, 1.862, 0.011
(node:78160) UnhandledPromiseRejectionWarning: Error: Failed to launch the browser process!


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md

    at onClose (/Users/damus/git/stmicro/es-theia/node_modules/puppeteer/lib/Launcher.js:750:14)
    at Interface.<anonymous> (/Users/damus/git/stmicro/es-theia/node_modules/puppeteer/lib/Launcher.js:739:50)
    at Interface.emit (events.js:326:22)
    at Interface.close (readline.js:416:8)
    at Socket.onend (readline.js:194:10)
    at Socket.emit (events.js:326:22)
    at endReadableNT (_stream_readable.js:1241:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
(node:78160) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:78160) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
"@theia/getting-started": "1.17.0", NaN, NaN, NaN, NaN
"@theia/git": "1.17.0", 2.154, 0.021, 0.975, 0.124
```

cdamus commented on package.json:
```diff
@@ -34,6 +34,7 @@
     "ignore-styles": "^5.0.1",
     "jsdom": "^11.5.1",
     "lerna": "^4.0.0",
+    "node-jq": "2.0.0",
```
Also, what is the necessity of adding this dependency? Or was it already indirectly required and is now being added here because its API is being used directly? Perhaps this can be summarized in the PR description.

sgraband replied:
Changed it to a caret version! The package is needed within the script to easily parse the `examples/browser/package.json` file and add dependencies programatically to it.

sgraband replied:
(Added a Note in the PR description: `Note: Added the node-jq dependency to the root package.json. It is used for json parsing in the script.`)

sgraband replied:
@cdamus I was able to remove the whole package and replace it with javascript native json parsing.

sgraband commented:
Thanks a lot for the feedback Christian! Regarding your comments:

I updated all of the inline comments you had.

To force the `examples/browser/package.json` to reset when the script is cancelled i added an exitHandler on `SIGINT` and `exit` that cleans the workspace. If we would do the whole calculation on a clean copy it would require a lot of steps to configure the copy into the environment (root package.json and tsconfig, etc.), which would then just alter different files, which will not be reset after exiting the script. And if we would contribute the copy of the browser-app, it would need to be maintained and also have the same problem, with unintended change being pushed. So i think this approach is preferable. The only downside i could find is, that this returns an error for the current method, but i don`t think this can be negated. WDYT?

Additionally, i added a `--url` parameter, which the user can use to define a different hostname than localhost:3000 and also to define a workspace via the url. I think this way we do not need to add 2 parameters (hostname and workspace) and tinker them together in the code to pass them to the `measure-performance.js` script.

Last but not least i updated the output of the script, so that the NaN errors are detected and replaced with an `Error while measuring with this extension` entry in the mean column. The other columns are decorated with a simple `-`.

cdamus commented on scripts/performance/extension-impact.js:
```diff
@@ -0,0 +1,177 @@
+            type: 'boolean'
+        }).option('url', {
+            alias: 'u',
+            desc: 'Specify a custom url to launch theia (e.g. with a specific workspace)',
```
URL should be capitalized.

cdamus commented on scripts/performance/README.md:
```diff
@@ -27,3 +27,40 @@ The script accepts the following optional parameters:
+    _For example: `--extensions '""@theia/git":"1.17.0""' '""@theia/keymaps":"1.17.0""'`_
+
+-   `--yarn`: Flag to trigger a full yarn at script startup (e.g. to build changes to extensions)
+-   `--url`: Specify a url that theia should be launched with (can also be used to specify the workspace to be opened) (default: `http://localhost:3000/#/<pathToMeasurementScript>/workspace`)
```
I really like the new URL parameter, thanks! But one minor point: URL is an initialism and should be written in all capitals.

sgraband replied:
Would you also expect the option itself to be capitalized?

cdamus commented on scripts/performance/extension-impact.js:
```diff
@@ -0,0 +1,177 @@
+})();
+
+async function extensionImpact(extensions) {
+    console.log(`Extension name, MEAN (${runs} runs) (in s), STDEV (in ms), CV (%), DIFF (in ms)`);
```
I'm sorry, but I should have noticed in my first review that I think the standard deviation and difference compared to the baseline are actually reported in seconds, not milliseconds as the column titles indicate.

Also, again I should have mentioned this the first time, the all-caps titles of the statistics columns are incongruent with the first column and look strange. I think better would just be normal title case: "Mean", "Std Dev", "CV" (this one is an initialism, so yes) and "Delta" (which I think is more the usual statistical term than diff, which has another specific meaning for programmers).

cdamus replied:
Oh, and yes, I should ask for "Extension Name" also in title case.

sgraband replied:
Thanks for the input! I changed the naming.

cdamus reviewed (commented):
Thanks for the changes! I particularly like the new URL parameter, which I've already put to use on an investigation of my own 😀.

The signal handler approach to clean-up is very tidy; I like it. About the error that you indicated inevitably happens, did you mean this?

```console
Extension name, MEAN (10 runs) (in s), STDEV (in ms), CV (%), DIFF (in ms)
^CBase theia, Error while measuring with this extension, -, -, -
internal/fs/utils.js:269
    throw err;
    ^

Error: ENOENT: no such file or directory, copyfile './backup-package.json' -> '../../examples/browser/package.json'
    at copyFileSync (fs.js:1907:3)
    at cleanWorkspace (/Users/damus/git/stmicro/es-theia/scripts/performance/extension-impact.js:117:5)
    at exitHandler (/Users/damus/git/stmicro/es-theia/scripts/performance/extension-impact.js:28:5)
    at process.emit (events.js:314:20)
    at process.exit (internal/process/per_thread.js:168:15)
    at exitHandler (/Users/damus/git/stmicro/es-theia/scripts/performance/extension-impact.js:29:13)
    at process.emit (events.js:314:20) {
  errno: -2,
  syscall: 'copyfile',
  code: 'ENOENT',
  path: './backup-package.json',
  dest: '../../examples/browser/package.json'
}
```

The "Error while measuring this extension" is quite innocuous and can be expected because, yes, I caused the measurement to fail. But the exception I think can be avoided by first checking whether the file is there to copy? Otherwise, the handler did actually work: cancelling leaves my Git checkout clean. So that's good!

sgraband commented:
I believe i solved the error you were experiencing. (At least i cannot reproduce it anymore. Let me know if you still can :smile:)

It seems that it was caused by ending the process to fast. I now made the cleanWorkspace() method async and await it in the handler and it seems to work.

sgraband commented:
(Pushed again to remove the yarn.lock changes that are no longer needed since the node-jq package is removed again.)

cdamus reviewed (changes requested):
Interesting that you removed the jq dependency: the code was a bit clearer to me using that API, but I can see that the use case was simple enough not to need it. So, that's fair.

Thanks also for capitalizing occurrences of "Theia". I wasn't sure about that because I thought maybe the lower-case NPM package name was the conventional usage.

It seems to me that the fix for the exception trace in the signal handler actually makes the problem worse: now the first ^C doesn't actually terminate the script. I have to press ^C again several times and then there's an unhandled promise rejection traced.

```console
$ node ./extension-impact.js --url "http://localhost:3000/#/Users/damus/git/Papyrus"                             
Extension Name, Mean (10 runs) (in s), Std Dev (in s), CV (%), Delta (in s)
Base Theia, 2.021, 0.026, 1.286, -
"@theia/bulk-edit": "1.17.0", 2.008, 0.010, 0.498, -0.013
^C^C^C^C^C^C^C"@theia/callhierarchy": "1.17.0", Error while measuring with this extension, -, -, -
^C(node:21552) UnhandledPromiseRejectionWarning: Error: Command failed: yarn browser build
    at checkExecSyncError (child_process.js:635:11)
    at execSync (child_process.js:671:15)
    at calculateExtension (/Users/damus/git/stmicro/es-theia/scripts/performance/extension-impact.js:151:9)
    at extensionImpact (/Users/damus/git/stmicro/es-theia/scripts/performance/extension-impact.js:100:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async /Users/damus/git/stmicro/es-theia/scripts/performance/extension-impact.js:84:5
```

sgraband commented:
Thanks! i will have a look at the error problem.

I pushed an additional commit (so the changes are easily readable; will squash if they are fine with you) that replaces the copying of the `base-package.json `to the `example/browser/package.json`. 
Removing the dependency led to some issues, where sometimes the copying was happening after we already added the new dependency. And since we already have the `base-package.json` as a json object i changed it so we make a copy of this, add the dependency and write the object to the file. This way we should not have any timing issues and we can keep out the dependency (which is preferable as we do not have to look after licenses etc.)

cdamus commented:
> I pushed an additional commit (so the changes are easily readable; will squash if they are fine with you) that replaces the copying of the `base-package.json `to the `example/browser/package.json`.

I don't know whether it's this specific change or something else, but with this latest commit, four out of five trials of running the test on all packages (no `--extensions` option specified) hung after measuring a few extensions. The fifth run completed, albeit still with a few unhandled promise exceptions that seem to be caused by the browser having been closed unexpectedly (no idea why that happens), and then another after that.

Is there something useful I can extract from a hung `extension-impact.js` process that could help to diagnose it?

> Removing the dependency led to some issues, where sometimes the copying was happening after we already added the new dependency. And since we already have the `base-package.json` as a json object i changed it so we make a copy of this, add the dependency and write the object to the file. This way we should not have any timing issues and we can keep out the dependency (which is preferable as we do not have to look after licenses etc.)

✅  Yeah, I see now that we basically make a clean copy of the original for every iteration of the test, modify it, and save.

sgraband commented:
> It seems to me that the fix for the exception trace in the signal handler actually makes the problem worse: now the first ^C doesn't actually terminate the script. I have to press ^C again several times and then there's an unhandled promise rejection traced.

Does it actually not terminate or do you just think it does not close quickly enough? For me it closes about a second after i press ^C, because it is waiting for the workspace cleaning. Also i believe pressing ^C multiple times still gives the same error, as just pressing it once and waiting right? So basically we get the same result, with the difference, that the user just gets impatient. Maybe i should add an `console.log('Exiting script: waiting for workspace cleanup')` before the clean up?

> I don't know whether it's this specific change or something else, but with this latest commit, four out of five trials of running the test on all packages (no --extensions option specified) hung after measuring a few extensions. The fifth run completed, albeit still with a few unhandled promise exceptions that seem to be caused by the browser having been closed unexpectedly (no idea why that happens), and then another after that.

Unfortunately i cannot reproduce this issue at all. I now ran three successful full runs of the script (without extensions defined) and none had any error. Did you specify any other parameter? And do you have the specific error message you get?

> Is there something useful I can extract from a hung extension-impact.js process that could help to diagnose it?

Currently not, what you could do is disable the filtering of all the console logs. For this you need to remove the `--log-level=fatal` from the `start` script in the `base-package.json` and replace the `stdio: 'pipe'` with `stdio: 'inherit'` in the script (There should be multiple). Additionally you could try to manually run the `measure-performance.js` script for the extension that it fails, to see if it can reproduce the problem.

sgraband commented on scripts/performance/extension-impact.js:
```diff
@@ -0,0 +1,173 @@
+const { copyFileSync, readdirSync, writeFileSync, unlinkSync } = require('fs');
+const mkdirp = require('mkdirp');
+const path = require('path');
+const env = Object.assign({}, process.env);
+env.PATH = path.resolve("../../node_modules/.bin") + path.delimiter + env.PATH;
+const basePackage = require(`./base-package.json`);
```
For @sdirix the script also ran through without any problems.

However, he noticed that the script only runs when you have concurrently installed globally. To fix this i added the `node_modules/.bin` folder to a copy of `process.env` and set this copy as the ENV for the concurrently call in line 150. Unfortunately we cannot easily make the concurrently call a npm script, as then the parameter passing would be quite tricky.

cdamus commented:
I've just realized that I ran some trial runs with the `--extensions` option in which I misnamed the extension that I was testing: `@theia-git/` instead of `@theia/git`. Using the latter I get performance numbers that I might expected, but using the former I get numbers only very slightly different to "Base Theia" without any indication that the extension under test actually doesn't exist. Perhaps some validation of the inputs is missing?

sgraband commented:
> I've just realized that I ran some trial runs with the --extensions option in which I misnamed the extension that I was testing: @theia-git/ instead of @theia/git. Using the latter I get performance numbers that I might expected, but using the former I get numbers only very slightly different to "Base Theia" without any indication that the extension under test actually doesn't exist. Perhaps some validation of the inputs is missing?

A pre-validation of the packages might be a little bit out of scope for now, as we cannot make any assumptions about the naming of a extension/dependency. However i added a try block around the build statement, that checks for errors during the build of the browser-app and does not measure on failures. Currently, the build script does not fail if it cannot resolve a dependency, but we think it should. So we will open a bug for that and then the script can detect wrong extensions.

> Perhaps a note on the console to say "Cancelled. Cleaning up (this may take a few seconds)" would be reassuring

I tried to achieve this, however this i not as straight forward as it seems. This might be a good improvement down the line, but for now it is out of scope. Especially, since the cleaning of the workspace works, so these improvements would only be cosmetic.

I also rebased the ancestor branch as well as this one to the current master.

cdamus commented on scripts/performance/extension-impact.js:
```diff
@@ -0,0 +1,178 @@
+    try {
+        execSync('yarn browser build', { cwd: '../../', stdio: 'pipe' });
+    } catch (error) {
+        console.log(`${extensionQualifier}, Error while buidling the package.json, -, -, -`);
```
```suggestion
        console.log(`${extensionQualifier}, Error while building the package.json, -, -, -`);
```

cdamus reviewed (approved):
LGTM! I expect the new build failure handling will solve my problem of the test appearing successful for non-existent extensions when the build actually fails on an unknown extension. Let's just hope that may happen soon.

sdirix commented:
Merged upstream with https://github.com/eclipse-theia/theia/pull/10192
